File: Praktikum 3/binary_search.py
import time
nums = [1, 6, 8, 9, 10, 13]
target = int(input())
mid = int(len(nums))
left = 0
right = len(nums) - 1

while right >= left:
    mid = (right + left)//2
    if target == nums[mid]:
        print(mid)
        break
    elif target < nums[mid]:
        right = mid - 1
    elif target > nums[mid]:
        left = mid + 1
if target < nums[right]: 
    if right == 0:
        print(0)
    else:
        print(right - 1)
elif target > nums[left]:
    print(left + 1)
else:
    print(right + 1)

File: Praktikum 3/BinarySearch.c
#include <stdio.h>

int binarySearch(int arr[], int left, int right, int target) {
    while (left <= right) {

        int mid = left + (right - left) / 2;
        if (arr[mid] == target) {
            return mid;
        }
        if (arr[mid] < target) {
            left = mid + 1;
        }
        else {
            right = mid - 1;
        }
    }
    return -1;
}

int main(){
    int n, k; scanf("%d", &n);
    int fin[100001];
    if(n == 1){
        int y; 
        scanf("%d", &y);
        scanf("%d", &k);
        int x;
        for(int i = 0; i < k; i++){
            scanf("%d", &x);
            if(x == y){
                fin[i] = 0;
            }
            else{
                fin[i] = -1;
            }
            //printf("#%d, ", fin[i]);
        }
        printf("[%d", fin[0]);
        for(int i = 1; i < k; i++){
            printf(", %d", fin[i]);
        }
        printf("]\n");
    }
    else{
        int a[100001];
        for(int i = 0; i < n; i++){
            scanf("%d", &a[i]);
        }
        
        scanf("%d", &k);
        int qq[100001]; 
        int sz = 1;
        for(int i = 0; i < k; i++){
            int x; int ans;
            scanf("%d", &x);
            ans = binarySearch(a, 0, n, x);
            
            int temp; int id = ans;
            while(a[ans] == a[id]){
                id--;
            }
            fin[i] = id + 1;
            //printf("#%d\n", fin[i]);
        
        }
        printf("[%d", fin[0]);
        for(int i = 1; i < k; i++){
            printf(", %d", fin[i]);
        }
        printf("]\n");    
    }
    
}